exchanges/binance/binance_websocket.go

```diff
@@ -337,6 +337,11 @@ func (b *Binance) wsHandleData(respRaw []byte) error {
 				b.Name,
 				err)
 		}
+		// maker placed buy order, so taker market sell
+		side := order.Buy
+		if t.Maker {
+			side = order.Sell
+		}
```

Comment:
Being a maker or taker doesn't necessarily give you direction of order for this trade; because your maker can be on the bid liquidity (buy limit order) and the taker will hit the bids (sell market order) with their base amount whereas a maker can be on the ask liquidity (sell limit order) then your taker will lift the asks (buy market order) with their quote amount. I think a better approach might be to update trade.Data to have a field IsTaker or IsMaker.

Reply:
I had the same response at first.
But that's because we've misnamed the field.
[As per API docs](https://developers.binance.com/docs/binance-spot-api-docs/web-socket-api/market-data-requests#recent-trades) the field `m` is actually `isBuyerMaker`.

Since this is a trade, a market event, we know that if the buyer is the maker, then the seller is the taker, and this is a market SELL. And visa versa.

I suggest we make this clearer:
```suggestion
		if t.IsBuyerMaker { // Seller is Taker
		    t.Side = order.Sell
		} else { // Buyer is Taker
		    t.Side = order.Buy
		}
```
Note: I'm also implicitly suggesting moving declaration of td := trade.Data{} higher up so that we don't need to allocate a `side` var just to put it into another var.

Aside: This must be one of the only usages of `Trade.Update`, which I'm hoping to eradicate.

Reply:
`buyer` is the maker, then the seller is the `taker` and vice versa. That's sound. But direction can still be ambiguous, I do see where you are coming from but if you are a buyer (BTC USDT e.g.) and you are a maker you have put a limit order on the bid side whereas as a taker you are taking liquidity from the asks, but from the buyers perspective you are still going long, you are still picking up BTC. Or maybe I am just really confused which could be the case 😆

Reply:
> if you are a buyer (BTC USDT e.g.) and you are a maker you have put a limit order on the bid side whereas as a taker you are taking liquidity from the asks, but from the buyers perspective you are still going long, you are still picking up BTC.

I realised I shouldn't be categorical about this, because I'm not 100% sure.
When talking about the market trade itself, we're asking which "side" it should be visualised as.
The reason I think of it as Taker is direction is because if the orderbook gets filled in one direction, e.g. Maker's Limit Bid/Buy orders are getting filled, then we'd expect to see "Last trades" all in Red as SELLs, so that the Last trade fee matches the chart movement.
Example here, where the green Taker Buys on Binance match the green candles:
![Screenshot 2025-01-21 at 10 28 15](https://github.com/user-attachments/assets/ed0e6885-de67-4538-b02d-25a0fd7cc0a1)

Reply:
Yeah that last trade makes what you suggest correct for visualisation to the exchange. 👍

Reply:
@Beadko GK's suggestion please 🙏

Reply:
Good to be back too! :)

I've moved the trades declaration up and updated the TradeStream to match Binance api